integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/schema/CarbonAlterTableColRenameDataTypeChangeCommand.scala

```diff
@@ -0,0 +1,345 @@
+      }
+      val operationContext = new OperationContext
+      val alterTableColRenameAndDataTypeChangePreEvent =
+        AlterTableColRenameAndDataTypeChangePreEvent(sparkSession, carbonTable,
+          alterTableColRenameAndDataTypeChangeModel)
+      OperationListenerBus.getInstance()
+        .fireEvent(alterTableColRenameAndDataTypeChangePreEvent, operationContext)
+      val newColumnName = alterTableColRenameAndDataTypeChangeModel.newColumnName
+      val oldColumnName = alterTableColRenameAndDataTypeChangeModel.columnName
+      val carbonColumns = carbonTable.getCreateOrderColumn(tableName).asScala.filter(!_.isInvisible)
+      if (!carbonColumns.exists(_.getColName.equalsIgnoreCase(oldColumnName))) {
+        throwMetadataException(dbName, tableName, s"Column does not exist: $oldColumnName")
+      }
+
+      val carbonColumn = carbonColumns.filter(_.getColName.equalsIgnoreCase(oldColumnName))
+      if (carbonColumn.size != 1) {
+        throwMetadataException(dbName, tableName, s"Invalid Column: $oldColumnName")
+      }
+      if (alterTableColRenameAndDataTypeChangeModel.isColumnRename) {
+        // check whether new column name is already an existing column name
+        if (carbonColumns.exists(_.getColName.equalsIgnoreCase(newColumnName))) {
+          throw new MalformedCarbonCommandException(s"Column Rename Operation failed. New " +
+                                                    s"column name $newColumnName already exists" +
+                                                    s" in table $tableName")
+        }
+
+        // if the datatype is source datatype, then it is just a column rename operation, else do
+        // the datatype validation for not source datatype
+        if (carbonColumn.head.getDataType.getName
+          .equalsIgnoreCase(alterTableColRenameAndDataTypeChangeModel.dataTypeInfo.dataType)) {
+          isColumnRenameOnly = true
+        } else {
+          isBothColRenameAndDataTypeChange = true
+        }
+
+        // if column rename operation is on partition column, then fail the rename operation
+        if (null != carbonTable.getPartitionInfo) {
+          val partitionColumns = carbonTable.getPartitionInfo.getColumnSchemaList
+          partitionColumns.asScala.foreach {
+            col =>
+              if (col.getColumnName.equalsIgnoreCase(oldColumnName)) {
+                throw new MalformedCarbonCommandException(
+                  s"Column Rename Operation failed. Renaming " +
+                  s"the partition column $newColumnName is not " +
+                  s"allowed")
+              }
+          }
+        }
+      } else {
+        isDataTypeChangeOnly = true
+      }
+      if (isBothColRenameAndDataTypeChange || isDataTypeChangeOnly) {
+        validateColumnDataType(alterTableColRenameAndDataTypeChangeModel.dataTypeInfo,
+          carbonColumn.head)
+      }
+      // read the latest schema file
+      val tableInfo: TableInfo =
```

Comment:
Why not use carbonTable.getTableInfo

Reply:
this is thrift tableInfo object as we will be changing the schema thrift, so this is used